S3/S3.py

```diff
@@ -870,43 +874,7 @@ def urlencode_string(self, string, urlencoding_mode = None):
             ## Don't do any pre-processing
             return string
 
-        encoded = ""
-        ## List of characters that must be escaped for S3
-        ## Haven't found this in any official docs
-        ## but my tests show it's more less correct.
-        ## If you start getting InvalidSignature errors
-        ## from S3 check the error headers returned
-        ## from S3 to see whether the list hasn't
-        ## changed.
-        for c in string:    # I'm not sure how to know in what encoding
-                    # 'object' is. Apparently "type(object)==str"
-                    # but the contents is a string of unicode
-                    # bytes, e.g. '\xc4\x8d\xc5\xafr\xc3\xa1k'
-                    # Don't know what it will do on non-utf8
-                    # systems.
-                    #           [hope that sounds reassuring ;-)]
-            o = ord(c)
-            if (o < 0x20 or o == 0x7f):
-                if urlencoding_mode == "fixbucket":
-                    encoded += "%%%02X" % o
-                else:
-                    error(u"Non-printable character 0x%02x in: %s" % (o, string))
-                    error(u"Please report it to s3tools-bugs@lists.sourceforge.net")
-                    encoded += replace_nonprintables(c)
-            elif (o == 0x20 or  # Space and below
-                o == 0x22 or    # "
-                o == 0x23 or    # #
-                o == 0x25 or    # % (escape character)
-                o == 0x26 or    # &
-                o == 0x2B or    # + (or it would become <space>)
-                o == 0x3C or    # <
-                o == 0x3E or    # >
-                o == 0x3F or    # ?
-                o == 0x60 or    # `
-                o >= 123):      # { and above, including >= 128 for UTF-8
-                encoded += "%%%02X" % o
-            else:
-                encoded += c
+        encoded = quote_plus(string, safe="~/")
```

Comment:
Would it make things clearer to have helper functions "decode_from_s3()"
and "encode_to_s3()" to ensure content in/out at that boundry is utf-8
encoded/decoded, rather than using the unicodise()/deunicodise() pair which
relies on system (console) encoding?  Also, when converting file names /
paths, we should be using sys.getdefaultencoding() /
sys.getfilesystemencoding() for the conversions.  With all that, ideally we
can drop Config.encoding which is more often than not
locale.getpreferredencoding() which again is a console encoding, not a file
system encoding.

I think we really need 3 sets of helper functions then.

encode_to_s3() - always uses UTF-8
decode_from_s3()

encode_to_console() - uses locale.getpreferredencoding()
decode_from_console() - likely only used for evaluating argv[].

encode_to_filesystem() - uses sys.getfilesystemencoding() or
getsysdefaultencoding()
decode_from_filesystem()






On Fri, Mar 13, 2015 at 9:15 AM, Florent Viard <notifications@github.com>
wrote:

> In S3/S3.py
> <https://github.com/s3tools/s3cmd/pull/495#discussion_r26388328>:
>
> > -                    encoded += replace_nonprintables(c)
> > -            elif (o == 0x20 or  # Space and below
> > -                o == 0x22 or    # "
> > -                o == 0x23 or    # #
> > -                o == 0x25 or    # % (escape character)
> > -                o == 0x26 or    # &
> > -                o == 0x2B or    # + (or it would become <space>)
> > -                o == 0x3C or    # <
> > -                o == 0x3E or    # >
> > -                o == 0x3F or    # ?
> > -                o == 0x60 or    # `
> > -                o >= 123):      # { and above, including >= 128 for UTF-8
> > -                encoded += "%%%02X" % o
> > -            else:
> > -                encoded += c
> > +        encoded = quote_plus(string, safe="~/")
>
> I also think so. That was the first fix I had to do to support all the
> oriental characters even when on UTF-8 system. Sorry to have not separated
> that in another commit for more clarity.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/s3tools/s3cmd/pull/495/files#r26388328>.
>